homestar-wasm/README.md

```diff
@@ -23,15 +23,1054 @@
+
+type IPLDMapAsWit union {
+  | Record {string:any}
+  | Variant {string:any}
+  | List TupleAsMap
+} representation kinded
+
+type WitAsIpldMap union {
+  | Record {string:any}
+  | Variant {string:any}
+  | List TupleAsMap
+} representation kinded
+```
+
+#### WIT Options
+
+This section outlines the translation process between [WIT option runtime values][wit-val]
+(of type `option`) and varying IPLD values. An [`option`][wit-option] can be interpreted
+as either a `Ipld::Null` or of any other IPLD value.
+
+* **IPLD to WIT Translation**
+
+  When a WIT function expects an `option` as input, an `Ipld::Null` value is
+  mapped to the `None`/`Unit` case for a WIT option. Otherwise, any other IPLD
+  value will be mapped to its matching WIT runtime value directly.
+
+  **Example**:
+
+  ```wit
+  export fn: func(a: option<s32>) -> option<s32>;
+  ```
+
+  * `Some` case:
+
+    - **Json Input**:
+
+      ```json
+      {
+        "args": [1]
+      }
+      ```
+
+  * `None` case:
+
+    - **Json Input**:
+
+      ```json
+      {
+        "args": [null]
+      }
+      ```
+
+  `1` is converted into an `Ipld::Integer`, which is then translated and
+  passed into `fn` as an integer argument (`s32`), as the `Some` case of the
+  option.
+
+  `null` is converted into an `Ipld::Null`, which is then translated and
+  passed into `fn` as a `None`/`Unit` case of the option (i.e. no value in WIT).
+
+  Essentially, you can view this as `Ipld::Any` being the `Some` case and
+  `Ipld::Null` being the `None` case.
+
+* **WIT to IPLD Translation**:
+
+  Conversely, when an `option` value is returned from a WIT function, it can be
+  translated back into an `Ipld::Null` value if it's the `None`/`Unit` case, or
+  any other IPLD value if it's the `Some` case.
+
+**IPLD Schema Definitions**:
+
+``` ipldsch
+type IpldAsWitOption union {
+  | Some any
+  | None
+} representation kinded
+
+type WitAsIpldOption union {
+  | Some any
+  | None
+} representation kinded
+```
+
+#### WIT Results
+
+This section outlines the translation process between [WIT result runtime values][wit-val]
+(of type `result`) and varying IPLD values. We treat result as Left/Right
+[either][either] types over an `Ipld::List` of 2 elements.
+
+A [`result`][wit-result] can be interpreted as one of these patterns:
+
+- `Ok` (with a payload)
+
+  * **IPLD to WIT Translation**
+
+    When a WIT function expects a `result` as input, an `Ipld::List` value can
+    be mapped to the `Ok` case of the `result` WIT runtime value, including
+    a payload.
+
+    **Example**:
+
+    ```wit
+    export fn: func(a: result<s32, string>) -> result<s32, string>;
+    ```
+
+   Given an JSON input for this function:
+
+      ```json
+      {
+        "args": [[47, null]]
+      }
+      ```
+
+      `[47, null]` is converted into an `Ipld::List`, which is then translated
+      and passed into `fn` as an `Ok` case of the `result` argument with a
+      payload of `47` matching the `s32` type on the left.
```

Comment:
Something is off with the formatting here. Don't see what it is immediately, but this all ends up in a code block together.

Reply:
will check

Reply:
Seems resolved now @bgins. Lmk.